src/app/upgrade.cpp

```diff
@@ -39,9 +39,33 @@
 #include "base/utils/fs.h"
 #include "base/utils/io.h"
 #include "base/utils/string.h"
+#include "base/utils/version.h"
+#include "base/version.h"
 
 namespace
 {
+    using AppVersion = Utils::Version<int, 4>;
+    const QString versionKey = QStringLiteral("ConfigVersion");
```

Comment:
It isn't actually a version of config itself since it doesn't follow config format changes (i.e. different qBittorrent versions can have exactly the same config format).
Why not call it "qBittorrentVersion" or even just "Version"?

Reply:
To be precise, I intend this field to record the last used qbt version. So maybe call it "LastVersion"?

Reply:
Really I would avoid rewriting the config file unless it's required so I would just use separate versioning.

Reply:
>Really I would avoid rewriting the config file unless it's required so I would just use separate versioning.

I suppose you are talking another approach? Won't it be harder to maintain? I'm not too impressed how the webapi versioning is maintained...

Also the current PR approach can show when the migration function was introduced (the qbt version number) and it may be helpful on deciding when to drop them..

Reply:
> I suppose you are talking another approach?

Yes. Just have separate version of "config format" (single integer) and increment it once format is changed (independently from qBittorrent releases). 
If you link it to the qBittorrent version, it can cause problems for those who use intermediate builds, not just official ones (we have many such users, including the developers themselves), if the format changes several times between neighboring releases. Otherwise, we will have to accumulate all such format changes and merge them immediately before the release.

Reply:
>If you link it to the qBittorrent version, it can cause problems for those who use intermediate builds, not just official ones (we have many such users, including the developers themselves), if the format changes several times between neighboring releases. 

It is not a problem only happening now, is it? Doesn't testers have their dispensable profile for testing already?
And I don't think your approach will handle it better: the version number is updated but you'll still need to worry about the affected migration functions, when to trigger and how it will be done from the last (or many) intermediate states.
IMO one shouldn't worry about the intermediate states, those are not the main audience (users) should normally see. What matters to them is the result.

>Just have separate version of "config format" (single integer) and increment it once format is changed (independently from qBittorrent releases).

IMO it is meaningless/impractical to have a versioning on config file "format". There are too many possibilities that could change the config format, e.g. GUI code adjustment, enum element addition/deletion, new key/value addition, etc. The worst part is: you can't remember to bump it correctly every time, can you? This is my main argument against it.

Maybe you misunderstood what I'm trying to solve here: I want to make sure the migration functions are triggered only when necessary and currently is to trigger them only when upgrading from an earlier release.

Reply:
> It is not a problem only happening now, is it?

Is it generally a bad idea to try to solve an "old" problem?

>Doesn't testers have their dispensable profile for testing already?

I was talking not only about testers, but also about those users who want to get new features/fixes before maintainer "matures" to produce another release (producing releases is definitely an "old" problem of qBittorrent).
But of course you have the right to ignore this minority, leaving them to take care of it themselves. I'm not blaming you. I just don't mind fixing it if there's a way that won't cause us too much trouble. That's why I'm trying to analyze it more deeply, involving you in it as well.

>And I don't think your approach will handle it better: the version number is updated but you'll still need to worry about the affected migration functions, when to trigger and how it will be done from the last (or many) intermediate states.

What scared you so much? This is supposed to work exactly the same as in your approach: when we need to update the format, we simply increase its number (change the constant in the code) and add update functions. The upgrade path looks the same as yours:
1. Read version from file
2. If read version is different than current one apply appropriate upgrade functions and rewrite version with the "current" one:
```
if (versionFromFile != currentVersion)
{
   if (versionFromFile < 2)
       // ...
   if (versionFromFile < 3)
       // ...
   if (versionFromFile < 4)
       // ...

    store(currentVersion);
}
```

Even talking about testers, it can add to their problems. Previously, they could use the same test profile with several (many) consecutive updates of the master branch with it. Now (using your approach) they will have to take extra care about changing the format (do not forget that master can have the same version for a very long time). To be honest, I can't even immediately guess how to handle it in a convenient way, because starting testing with a clean profile every time is not always an acceptable solution).

>There are too many possibilities that could change the config format, e.g. GUI code adjustment, enum element addition/deletion, new key/value addition, etc. 

Do you care about all this here? After all, we are talking only about those changes that require an explicit upgrade. We consider everything else to be the same format.

>I want to make sure the migration functions are triggered only when necessary and currently is to trigger them only when upgrading from an earlier release.

I understand and approve of your goal. But I have questions about the way to achieve it, which I outlined above. I'm still sure it's going to give us an unnecessary pain in the ass. I'm not pushing for another solution if I'm wrong about yours.

Reply:
>To be honest, I can't even immediately guess how to handle it in a convenient way, because starting testing with a clean profile every time is not always an acceptable solution).

Well, removing version from existing config is probably enough.
Wouldn't it be better if alpha/beta/rc versions don't update `qBittorrentVersion` in config?

Reply:
>Is it generally a bad idea to try to solve an "old" problem?

No, except that I don't really consider it (intermediate states) a must-be-solved problem.

>I was talking not only about testers, but also about those users who want to get new features/fixes before maintainer "matures" to produce another release (producing releases is definitely an "old" problem of qBittorrent).

Those are nightly builds and everyone understand the risks of using them.

>What scared you so much? This is supposed to work exactly the same as in your approach

You missed the point. I'm trying to say that the "intermediate states" problem still manifests in your approach, it isn't handled any better.

>Wouldn't it be better if alpha/beta/rc versions don't update qBittorrentVersion in config?

They don't. I'm seeing `qBittorrentVersion == 4.4.0.0` in config now and this is v4.4.0RC1 in about dialog. Does it help?

Sorry, I still think there are some misunderstanding here:
1. I consider your suggested approach is very similar to mine with the only exception that you must bump the version number manually. And I consider it a minor maintenance shortcoming on your approach.
2. I still don't understand your worries. For the problems you mentioned, either they aren't handled better in you approach or they aren't really a problem, or maybe I don't really understand what you meant...

Reply:
> I'm trying to say that the "intermediate states" problem still manifests in your approach, it isn't handled any better.

Then I don't understand what exactly you mean by "intermediate states" so that "it isn't any handled better".

>Those are nightly builds and everyone understand the risks of using them.

Given qBittorrent's problems with issuing official releases, many so-called "nightly" builds are more stable than official release builds, so I don't see any real reason for your dismissive attitude towards them.
But even if you discount this and stick to pure formalism in the issue of releases, your approach still introduces new problems to developers and testers. At least for me it looks much easier and more reliable to increase the version number exactly when it is required than to suggest to several people (other developers and testers) to be constantly on the alert and track the fact of the configuration upgrade in order to manually apply some actions to their profiles in this regard.
>They don't. I'm seeing `qBittorrentVersion == 4.4.0.0` in config now and this is v4.4.0RC1 in about dialog. Does it help?

No. I mean that it shouldn't (re)write `qBittorrentVersion`.
Let's look at the development workflow starting, for example, with the release of v4.4. As soon as it is released, the master branch will be assigned version 4.5.0alpha and will remain so for almost the entire 4.4.x release cycle, so as soon as you launch it for the first time, it will write `qBittorrentVersion=4.5.0.0` to the configuration file. Now, let's say you add some upgrade function for v4.4.1. It will not be applied unless you delete the `qBittorrentVersion` from the file.

Reply:
>Given qBittorrent's problems with issuing official releases, many so-called "nightly" builds are more stable than official release builds, so I don't see any real reason for your dismissive attitude towards them.

Your are mistaken, it is not "dismissive"... If you want to promote nightly builds then there is no point of having an exact official release, you just let people download whatever you subjectively(?) deemed stable build. And this is the topic of release engineering, I don't want to see this PR bogged down with it...

BTW your attitude here is inconsistent with the webAPI version bump. IIRC you only permit one bump per release, do you intend to relax this restriction from now on? because webapi users on nightly builds experience the same (conceptually) too.

>Let's look at the development workflow starting, for example, with the release of v4.4. As soon as it is released, the master branch will be assigned version 4.5.0alpha and will remain so for almost the entire 4.4.x release cycle, so as soon as you launch it for the first time, it will write qBittorrentVersion=4.5.0.0 to the configuration file. Now, let's say you add some upgrade function for v4.4.1. It will not be applied unless you delete the qBittorrentVersion from the file.

It is a good example but I have some further questions before setting my choice. How would you maintain the `configVersion` in your approach?
Lets say master branch now start with `configVersion = 1`. Later someone added some migration code and it becomes to `configVersion = 3`. v4.4.0 is branched out earlier from master branch and it starts with `configVersion = 1` but for v4.4.1 there is a unique migration code (not needed in master branch) for it and what is the `configVersion` for the unique migration code?
`2` is not a good choice because master branch has already taken it.
`4` is feasible but then master branch will be forced to bump it even if there is no change in code. This implies we must maintain `configVersion` uniqueness across all git branches.
Does the above sounds reasonable?

Reply:
>BTW your attitude here is inconsistent with the webAPI version bump. IIRC you only permit one bump per release, do you intend to relax this restriction from now on? because webapi users on nightly builds experience the same (conceptually) too.

Since I've finally become disillusioned with our release cycle, I'd rather relax this rule for webAPI.
Or we still have to change the release production policy in order to be more flexible and faster if necessary. But this is @sledgehammer999 domain, so we have no power to change it ourselves.

Reply:
>but for v4.4.1 there is a unique migration code (not needed in master branch) for it

Can't imagine what unique migration do you mean. 
Usually we don't develop a release branch so different from the master. It almost completely follows the master until there are such changes that we do not want to add to the release branch. But since then, the release branch is supposed to accept only minimal changes that should not affect the config file format, at least in some unique way.

Reply:
One more remark. It looks like the "config version" is an incorrect name for my approach. It cannot be guaranteed to mean that the file has the appropriate format, even if the file is modified only by qBittorrent itself. Your approach has the same drawback. It seems that after the downgrade we will have a downgraded version recorded in the file, but some settings may remain in a newer format. Not to mention that the config can be changed manually... In general, this record can only mean a declaration like "this file should be treated as having this format" or even "this file may contain options in this and newer format".

Reply:
PR updated, also added some other minor cleanups.